File: showmecode_100/java/com/interview/array/ReorderArrayByIndex.java
package com.interview.array;

import java.util.Arrays;

/**
 * Date 12/29/2015
 * @author Tushar Roy
 *
 * Given two arrays one with values and other with index where values should be positioned. Move values to correct
 * position
 *
 * Time complexity - O(n)
 * Space complexity - O(1)
 *
 * http://www.geeksforgeeks.org/reorder-a-array-according-to-given-indexes/
 */
public class ReorderArrayByIndex {
    public void reorder(int input[], int index[]) {
        if(index.length != input.length) {
            throw new IllegalArgumentException();
        }
        for (int i = 0 ; i < index.length; i++) {
            while (index[i] != i) {
                int sIndex = index[index[i]];
                int sVal = input[index[i]];

                index[index[i]] = index[i];
                input[index[i]] = input[i];

                index[i] = sIndex;
                input[i] = sVal;
            }
        }
    }

    public static void main(String args[]) {
        int input[] = {50, 40, 70, 60, 90};
        int index[] = {3,  0,  4,  1,  2};
        ReorderArrayByIndex reorderArrayByIndex = new ReorderArrayByIndex();
        reorderArrayByIndex.reorder(input, index);
        Arrays.stream(input).forEach(i -> System.out.print(i + " "));
        System.out.println();
        Arrays.stream(index).forEach(i -> System.out.print(i + " "));
    }
}

File: showmecode_100/python/array/reorderarraybyindex.py
# http://www.geeksforgeeks.org/reorder-a-array-according-to-given-indexes/

def reorder(input, index):
    if len(input) != len(index):
        raise ValueError
    for i in range(len(index)):
        while index[i] != i:
            s_index = index[index[i]]
            s_val = input[index[i]]

            index[index[i]] = index[i]
            input[index[i]] = input[i]

            index[i] = s_index
            input[i] = s_val

if __name__ == '__main__':
    input = [50, 40, 70, 60, 90]
    index = [3, 0, 4, 1, 2]

    reorder(input, index)
    print(input)
    print(index)

            
